crates/red_knot_python_semantic/src/types/type_ordering.rs

```diff
@@ -261,11 +266,38 @@ pub(super) fn union_elements_ordering<'db>(left: &Type<'db>, right: &Type<'db>)
         (Type::Dynamic(_), _) => Ordering::Less,
         (_, Type::Dynamic(_)) => Ordering::Greater,
 
-        (Type::Union(left), Type::Union(right)) => left.cmp(right),
+        (Type::Union(left), Type::Union(right)) => {
+            // Lexicographically compare the elements of the two unions.
+            for (left, right) in left.elements(db).iter().zip(right.elements(db)) {
+                let ordering = union_elements_ordering(left, right, db);
+                if ordering != Ordering::Equal {
+                    return ordering;
+                }
+            }
+            left.elements(db).len().cmp(&right.elements(db).len())
```

Comment:
Would it be more efficient to do this check first, to avoid lots of element comparisons if we compare two lengthy unions of different lengths?